File: apps/backend/src/services/auth.service.ts
import jwt from 'jsonwebtoken';
import prisma from '../lib/prisma.js';
import redisClient from '../lib/redis.js';
import { sms } from '../lib/africastalking.js';
import { 
  trackFailedLogin, 
  clearFailedLoginAttempts, 
  isAccountLocked,
  detectAndHandleBreach 
} from './security.service.js';

const JWT_SECRET = process.env.JWT_SECRET || 'your-secret-key-change-in-production';
const JWT_REFRESH_SECRET = process.env.JWT_REFRESH_SECRET || 'your-refresh-secret-key';
const JWT_EXPIRES_IN = '24h';
const REFRESH_TOKEN_EXPIRES_IN = '30d';
const OTP_EXPIRY_SECONDS = 300; // 5 minutes

interface OTPData {
  code: string;
  phoneNumber: string;
  attempts: number;
  createdAt: number;
}

/**
 * Generate a 6-digit OTP code
 */
export const generateOTP = (): string => {
  return Math.floor(100000 + Math.random() * 900000).toString();
};

/**
 * Generate a unique referral code
 */
export const generateReferralCode = async (): Promise<string> => {
  let code: string;
  let exists = true;

  while (exists) {
    // Generate 8-character alphanumeric code
    code = Math.random().toString(36).substring(2, 10).toUpperCase();
    const user = await prisma.user.findUnique({
      where: { referralCode: code },
    });
    exists = !!user;
  }

  return code!;
};

/**
 * Send OTP via SMS using Africa's Talking
 */
export const sendOTP = async (phoneNumber: string): Promise<{ success: boolean; message: string }> => {
  try {
    // Check rate limiting: 3 OTP requests per 10 minutes
    const rateLimitKey = `otp:ratelimit:${phoneNumber}`;
    const requestCount = await redisClient.get(rateLimitKey);
    
    if (requestCount && parseInt(requestCount) >= 3) {
      return {
        success: false,
        message: 'Too many OTP requests. Please try again in 10 minutes.',
      };
    }

    // Generate OTP
    const code = generateOTP();
    
    // Store OTP in Redis with 5-minute expiry
    const otpKey = `otp:${phoneNumber}`;
    const otpData: OTPData = {
      code,
      phoneNumber,
      attempts: 0,
      createdAt: Date.now(),
    };
    
    await redisClient.setEx(otpKey, OTP_EXPIRY_SECONDS, JSON.stringify(otpData));
    
    // Increment rate limit counter
    const currentCount = await redisClient.incr(rateLimitKey);
    if (currentCount === 1) {
      await redisClient.expire(rateLimitKey, 600); // 10 minutes
    }

    // Send SMS
    const message = `Your Tikit verification code is: ${code}. Valid for 5 minutes.`;
    
    // Only send SMS if not in test mode
    if (process.env.NODE_ENV !== 'test') {
      await sms.send({
        to: [phoneNumber],
        message,
        from: process.env.AFRICASTALKING_SENDER_ID || 'Tikit',
      });
    }

    return {
      success: true,
      message: 'OTP sent successfully',
    };
  } catch (error) {
    console.error('Error sending OTP:', error);
    return {
      success: false,
      message: 'Failed to send OTP. Please try again.',
    };
  }
};

/**
 * Verify OTP code
 */
export const verifyOTP = async (
  phoneNumber: string,
  code: string
): Promise<{ success: boolean; message: string }> => {
  try {
    const otpKey = `otp:${phoneNumber}`;
    const otpDataStr = await redisClient.get(otpKey);

    if (!otpDataStr) {
      return {
        success: false,
        message: 'OTP expired or not found. Please request a new one.',
      };
    }

    const otpData: OTPData = JSON.parse(otpDataStr);

    // Check for brute force: 5 failed attempts locks for 30 minutes
    if (otpData.attempts >= 5) {
      const lockKey = `otp:lock:${phoneNumber}`;
      await redisClient.setEx(lockKey, 1800, 'locked'); // 30 minutes
      await redisClient.del(otpKey);
      
      // Find user and lock their account if they exist
      const user = await prisma.user.findUnique({
        where: { phoneNumber },
      });
      
      if (user) {
        await detectAndHandleBreach(
          user.id,
          'failed_login',
          { reason: 'Too many failed OTP attempts' }
        );
      }
      
      return {
        success: false,
        message: 'Too many failed attempts. Account locked for 30 minutes.',
      };
    }

    // Verify code
    if (otpData.code !== code) {
      otpData.attempts += 1;
      await redisClient.setEx(otpKey, OTP_EXPIRY_SECONDS, JSON.stringify(otpData));
      
      return {
        success: false,
        message: `Invalid OTP code. ${5 - otpData.attempts} attempts remaining.`,
      };
    }

    // OTP verified successfully - delete it
    await redisClient.del(otpKey);

    return {
      success: true,
      message: 'OTP verified successfully',
    };
  } catch (error) {
    console.error('Error verifying OTP:', error);
    return {
      success: false,
      message: 'Failed to verify OTP. Please try again.',
    };
  }
};

/**
 * Check if phone number is locked due to failed attempts
 */
export const isPhoneLocked = async (phoneNumber: string): Promise<boolean> => {
  const lockKey = `otp:lock:${phoneNumber}`;
  const locked = await redisClient.get(lockKey);
  return !!locked;
};

/**
 * Register a new user
 */
export const registerUser = async (data: {
  phoneNumber: string;
  firstName?: string;
  lastName?: string;
  email?: string;
  state: string;
  preferredLanguage?: string;
  referredBy?: string;
  role?: string;
  organizationName?: string;
  organizationType?: string;
}): Promise<{ success: boolean; user?: any; message: string }> => {
  try {
    // Check if user already exists
    const existingUser = await prisma.user.findUnique({
      where: { phoneNumber: data.phoneNumber },
    });

    if (existingUser) {
      return {
        success: false,
        message: 'User with this phone number already exists',
      };
    }

    // Validate role
    const role = data.role || 'attendee';
    if (!['attendee', 'organizer'].includes(role)) {
      return {
        success: false,
        message: 'Invalid role. Must be attendee or organizer',
      };
    }

    // Generate unique referral code
    const referralCode = await generateReferralCode();

    // Create user
    const user = await prisma.user.create({
      data: {
        phoneNumber: data.phoneNumber,
        phoneVerified: true, // Already verified via OTP
        firstName: data.firstName,
        lastName: data.lastName,
        email: data.email,
        state: data.state,
        preferredLanguage: data.preferredLanguage || 'en',
        referralCode,
        referredBy: data.referredBy,
        role,
        organizationName: data.organizationName,
        organizationType: data.organizationType,
      },
    });

    // If referred by someone, create referral record
    if (data.referredBy) {
      const referrer = await prisma.user.findUnique({
        where: { referralCode: data.referredBy },
      });

      if (referrer) {
        await prisma.referral.create({
          data: {
            referrerId: referrer.id,
            referredUserId: user.id,
            status: 'pending',
            rewardAmount: 200, // ₦200 reward
          },
        });
      }
    }

    return {
      success: true,
      user: {
        id: user.id,
        phoneNumber: user.phoneNumber,
        firstName: user.firstName,
        lastName: user.lastName,
        email: user.email,
        state: user.state,
        preferredLanguage: user.preferredLanguage,
        role: user.role,
        walletBalance: user.walletBalance || 0,
        referralCode: user.referralCode,
        organizationName: user.organizationName,
        organizationType: user.organizationType,
        isVerified: user.isVerified || false,
        createdAt: user.createdAt?.toISOString() || new Date().toISOString(),
      },
      message: 'User registered successfully',
    };
  } catch (error) {
    console.error('Error registering user:', error);
    return {
      success: false,
      message: 'Failed to register user. Please try again.',
    };
  }
};

/**
 * Generate JWT access token
 */
export const generateAccessToken = (userId: string, role: string, state: string): string => {
  return jwt.sign({ userId, role, state }, JWT_SECRET, {
    expiresIn: JWT_EXPIRES_IN,
  });
};

/**
 * Generate JWT refresh token
 */
export const generateRefreshToken = (userId: string): string => {
  return jwt.sign({ userId }, JWT_REFRESH_SECRET, {
    expiresIn: REFRESH_TOKEN_EXPIRES_IN,
  });
};

/**
 * Verify JWT access token
 */
export const verifyAccessToken = (token: string): { userId: string; role: string; state: string } | null => {
  try {
    const decoded = jwt.verify(token, JWT_SECRET) as { userId: string; role: string; state: string };
    return decoded;
  } catch (error) {
    return null;
  }
};

/**
 * Verify JWT refresh token
 */
export const verifyRefreshToken = (token: string): { userId: string } | null => {
  try {
    const decoded = jwt.verify(token, JWT_REFRESH_SECRET) as { userId: string };
    return decoded;
  } catch (error) {
    return null;
  }
};

/**
 * Login user
 */
export const loginUser = async (
  phoneNumber: string,
  ipAddress?: string,
  userAgent?: string
): Promise<{ success: boolean; user?: any; accessToken?: string; refreshToken?: string; message: string }> => {
  try {
    const user = await prisma.user.findUnique({
      where: { phoneNumber },
    });

    if (!user) {
      // Track failed login attempt
      const { shouldLock, attemptsRemaining } = await trackFailedLogin(
        phoneNumber,
        ipAddress,
        userAgent
      );
      
      if (shouldLock) {
        return {
          success: false,
          message: 'Too many failed login attempts. Please try again in 30 minutes.',
        };
      }
      
      return {
        success: false,
        message: `User not found. ${attemptsRemaining} attempts remaining before account lock.`,
      };
    }

    // Check if account is locked
    const lockStatus = await isAccountLocked(user.id);
    if (lockStatus.locked) {
      const expiresIn = lockStatus.expiresAt 
        ? Math.ceil((lockStatus.expiresAt - Date.now()) / 60000) 
        : 30;
      
      return {
        success: false,
        message: `Account is locked due to ${lockStatus.reason}. Please try again in ${expiresIn} minutes.`,
      };
    }

    // Clear any failed login attempts on successful login
    await clearFailedLoginAttempts(phoneNumber);

    // Generate tokens
    const accessToken = generateAccessToken(user.id, user.role, user.state);
    const refreshToken = generateRefreshToken(user.id);

    // Store refresh token in Redis
    const refreshTokenKey = `refresh:${user.id}`;
    await redisClient.setEx(refreshTokenKey, 30 * 24 * 60 * 60, refreshToken); // 30 days

    return {
      success: true,
      user: {
        id: user.id,
        phoneNumber: user.phoneNumber,
        firstName: user.firstName,
        lastName: user.lastName,
        email: user.email,
        state: user.state,
        preferredLanguage: user.preferredLanguage,
        role: user.role,
        walletBalance: user.walletBalance || 0,
        referralCode: user.referralCode || '',
        organizationName: user.organizationName,
        organizationType: user.organizationType,
        isVerified: user.isVerified || false,
        createdAt: user.createdAt?.toISOString() || new Date().toISOString(),
      },
      accessToken,
      refreshToken,
      message: 'Login successful',
    };
  } catch (error) {
    console.error('Error logging in user:', error);
    return {
      success: false,
      message: 'Failed to login. Please try again.',
    };
  }
};

/**
 * Refresh access token
 */
export const refreshAccessToken = async (
  refreshToken: string
): Promise<{ success: boolean; accessToken?: string; message: string }> => {
  try {
    const decoded = verifyRefreshToken(refreshToken);
    
    if (!decoded) {
      return {
        success: false,
        message: 'Invalid refresh token',
      };
    }

    // Check if refresh token exists in Redis
    const refreshTokenKey = `refresh:${decoded.userId}`;
    const storedToken = await redisClient.get(refreshTokenKey);

    if (storedToken !== refreshToken) {
      return {
        success: false,
        message: 'Refresh token not found or expired',
      };
    }

    // Get user
    const user = await prisma.user.findUnique({
      where: { id: decoded.userId },
    });

    if (!user) {
      return {
        success: false,
        message: 'User not found',
      };
    }

    // Generate new access token
    const accessToken = generateAccessToken(user.id, user.role, user.state);

    return {
      success: true,
      accessToken,
      message: 'Token refreshed successfully',
    };
  } catch (error) {
    console.error('Error refreshing token:', error);
    return {
      success: false,
      message: 'Failed to refresh token. Please login again.',
    };
  }
};

/**
 * Logout user
 */
export const logoutUser = async (userId: string): Promise<{ success: boolean; message: string }> => {
  try {
    // Delete refresh token from Redis
    const refreshTokenKey = `refresh:${userId}`;
    await redisClient.del(refreshTokenKey);

    return {
      success: true,
      message: 'Logout successful',
    };
  } catch (error) {
    console.error('Error logging out user:', error);
    return {
      success: false,
      message: 'Failed to logout. Please try again.',
    };
  }
};

File: apps/backend-fastapi/services/auth_service.py
"""
Authentication service for user management
"""
from passlib.context import CryptContext
from jose import JWTError, jwt
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
import secrets
import string
from database import supabase_client
from config import settings
import logging

logger = logging.getLogger(__name__)

# Password hashing
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

class AuthService:
    def __init__(self):
        self.supabase = supabase_client.get_service_client()
        if not self.supabase:
            logger.warning("Supabase not configured, auth service will use mock responses")
    
    def hash_password(self, password: str) -> str:
        """Hash a password"""
        return pwd_context.hash(password)
    
    def verify_password(self, plain_password: str, hashed_password: str) -> bool:
        """Verify a password against its hash"""
        return pwd_context.verify(plain_password, hashed_password)
    
    def create_access_token(self, data: dict) -> str:
        """Create JWT access token"""
        to_encode = data.copy()
        expire = datetime.utcnow() + timedelta(minutes=settings.access_token_expire_minutes)
        to_encode.update({"exp": expire})
        return jwt.encode(to_encode, settings.jwt_secret, algorithm=settings.jwt_algorithm)
    
    def create_refresh_token(self, data: dict) -> str:
        """Create JWT refresh token"""
        to_encode = data.copy()
        expire = datetime.utcnow() + timedelta(days=settings.refresh_token_expire_days)
        to_encode.update({"exp": expire})
        return jwt.encode(to_encode, settings.jwt_refresh_secret, algorithm=settings.jwt_algorithm)
    
    def verify_token(self, token: str, secret: str) -> Optional[Dict[str, Any]]:
        """Verify JWT token"""
        try:
            payload = jwt.decode(token, secret, algorithms=[settings.jwt_algorithm])
            return payload
        except JWTError:
            return None
    
    def generate_referral_code(self) -> str:
        """Generate unique referral code"""
        return ''.join(secrets.choice(string.ascii_uppercase + string.digits) for _ in range(8))
    
    def normalize_phone_number(self, phone: str) -> str:
        """Normalize phone number to +234 format"""
        if phone.startswith('+234'):
            return phone
        elif phone.startswith('234'):
            return f'+{phone}'
        elif phone.startswith('0'):
            return f'+234{phone[1:]}'
        return phone
    
    async def register_user(self, user_data: dict) -> Dict[str, Any]:
        """Register a new user"""
        try:
            # Normalize phone number
            phone_number = self.normalize_phone_number(user_data['phone_number'])
            
            # Check if user already exists
            existing_user = self.supabase.table('users').select('id').eq('phone_number', phone_number).execute()
            if existing_user.data:
                return {
                    'success': False,
                    'error': {
                        'code': 'USER_EXISTS',
                        'message': 'Phone number already registered'
                    }
                }
            
            # Check email if provided
            if user_data.get('email'):
                existing_email = self.supabase.table('users').select('id').eq('email', user_data['email']).execute()
                if existing_email.data:
                    return {
                        'success': False,
                        'error': {
                            'code': 'EMAIL_EXISTS',
                            'message': 'Email address already registered'
                        }
                    }
            
            # Validate required role field
            if not user_data.get('role'):
                return {
                    'success': False,
                    'error': {
                        'code': 'MISSING_ROLE',
                        'message': 'User role is required'
                    }
                }
            
            # Validate role value
            valid_roles = ['attendee', 'organizer']
            if user_data['role'] not in valid_roles:
                return {
                    'success': False,
                    'error': {
                        'code': 'INVALID_ROLE',
                        'message': f'Role must be one of: {", ".join(valid_roles)}'
                    }
                }
            
            # Hash password
            hashed_password = self.hash_password(user_data['password'])
            
            # Generate referral code
            referral_code = self.generate_referral_code()
            
            # Create user record
            user_record = {
                'phone_number': phone_number,
                'password': hashed_password,
                'phone_verified': True,  # Auto-verify for now
                'first_name': user_data['first_name'],
                'last_name': user_data['last_name'],
                'email': user_data.get('email'),
                'state': user_data['state'],
                'preferred_language': user_data.get('preferred_language', 'en'),
                'role': user_data['role'],  # Use direct access since we validated it exists
                'organization_name': user_data.get('organization_name'),
                'organization_type': user_data.get('organization_type'),
                'referral_code': referral_code,
                'wallet_balance': 0.0,
                'is_verified': False,
                'created_at': datetime.utcnow().isoformat(),
                'updated_at': datetime.utcnow().isoformat()
            }
            
            # Insert user
            result = self.supabase.table('users').insert(user_record).execute()
            
            if not result.data:
                return {
                    'success': False,
                    'error': {
                        'code': 'INTERNAL_ERROR',
                        'message': 'Failed to create user'
                    }
                }
            
            user = result.data[0]
            
            # Generate tokens
            token_data = {
                'user_id': user['id'],
                'phone_number': user['phone_number'],
                'role': user['role'],
                'state': user['state']
            }
            
            access_token = self.create_access_token(token_data)
            refresh_token = self.create_refresh_token({'user_id': user['id']})
            
            return {
                'success': True,
                'message': 'Registration successful',
                'data': {
                    'user': {
                        'id': user['id'],
                        'phone_number': user['phone_number'],
                        'first_name': user['first_name'],
                        'last_name': user['last_name'],
                        'email': user.get('email'),
                        'state': user['state'],
                        'role': user['role'],
                        'wallet_balance': user['wallet_balance'],
                        'referral_code': user['referral_code'],
                        'organization_name': user.get('organization_name'),
                        'organization_type': user.get('organization_type'),
                        'is_verified': user['is_verified'],
                        'created_at': user['created_at']
                    },
                    'access_token': access_token,
                    'refresh_token': refresh_token
                }
            }
            
        except Exception as e:
            logger.error(f"Registration error: {e}")
            return {
                'success': False,
                'error': {
                    'code': 'INTERNAL_ERROR',
                    'message': 'Failed to register user'
                }
            }
    
    async def login_user(self, phone_number: str, password: str) -> Dict[str, Any]:
        """Login user with phone and password"""
        try:
            # Normalize phone number
            normalized_phone = self.normalize_phone_number(phone_number)
            
            # Find user
            result = self.supabase.table('users').select('*').eq('phone_number', normalized_phone).execute()
            
            if not result.data:
                return {
                    'success': False,
                    'error': {
                        'code': 'INVALID_CREDENTIALS',
                        'message': 'Invalid phone number or password'
                    }
                }
            
            user = result.data[0]
            
            # Verify password
            if not user.get('password') or not self.verify_password(password, user['password']):
                return {
                    'success': False,
                    'error': {
                        'code': 'INVALID_CREDENTIALS',
                        'message': 'Invalid phone number or password'
                    }
                }
            
            # Generate tokens
            token_data = {
                'user_id': user['id'],
                'phone_number': user['phone_number'],
                'role': user['role'],
                'state': user['state']
            }
            
            access_token = self.create_access_token(token_data)
            refresh_token = self.create_refresh_token({'user_id': user['id']})
            
            return {
                'success': True,
                'message': 'Login successful',
                'data': {
                    'user': {
                        'id': user['id'],
                        'phone_number': user['phone_number'],
                        'first_name': user['first_name'],
                        'last_name': user['last_name'],
                        'email': user.get('email'),
                        'state': user['state'],
                        'role': user['role'],
                        'wallet_balance': user.get('wallet_balance', 0.0),
                        'referral_code': user.get('referral_code', ''),
                        'organization_name': user.get('organization_name'),
                        'organization_type': user.get('organization_type'),
                        'is_verified': user.get('is_verified', False),
                        'created_at': user['created_at']
                    },
                    'access_token': access_token,
                    'refresh_token': refresh_token
                }
            }
            
        except Exception as e:
            logger.error(f"Login error: {e}")
            return {
                'success': False,
                'error': {
                    'code': 'INTERNAL_ERROR',
                    'message': 'Failed to login'
                }
            }
    
    async def get_user_by_id(self, user_id: str) -> Optional[Dict[str, Any]]:
        """Get user by ID"""
        try:
            result = self.supabase.table('users').select('*').eq('id', user_id).execute()
            return result.data[0] if result.data else None
        except Exception as e:
            logger.error(f"Get user error: {e}")
            return None
    
    async def refresh_access_token(self, refresh_token: str) -> Dict[str, Any]:
        """Refresh access token"""
        try:
            # Verify refresh token
            payload = self.verify_token(refresh_token, settings.jwt_refresh_secret)
            if not payload:
                return {
                    'success': False,
                    'error': {
                        'code': 'INVALID_TOKEN',
                        'message': 'Invalid refresh token'
                    }
                }
            
            # Get user
            user = await self.get_user_by_id(payload['user_id'])
            if not user:
                return {
                    'success': False,
                    'error': {
                        'code': 'USER_NOT_FOUND',
                        'message': 'User not found'
                    }
                }
            
            # Generate new access token
            token_data = {
                'user_id': user['id'],
                'phone_number': user['phone_number'],
                'role': user['role'],
                'state': user['state']
            }
            
            access_token = self.create_access_token(token_data)
            
            return {
                'success': True,
                'message': 'Token refreshed successfully',
                'access_token': access_token
            }
            
        except Exception as e:
            logger.error(f"Refresh token error: {e}")
            return {
                'success': False,
                'error': {
                    'code': 'INTERNAL_ERROR',
                    'message': 'Failed to refresh token'
                }
            }

# Global auth service instance
auth_service = AuthService()